vncviewer/XInputTouchHandler.cxx

```diff
@@ -0,0 +1,493 @@
+{
+  Window root, child;
+  int x_root, y_root;
+  XkbStateRec state;
+
+  // We don't have a real event to steal things from, so we'll have
+  // to fake these events based on the current state of things
+
+  root = XDefaultRootWindow(fl_display);
+  XTranslateCoordinates(fl_display, wnd, root,
+                        src.event_x,
+                        src.event_y,
+                        &x_root, &y_root, &child);
+  XkbGetState(fl_display, XkbUseCoreKbd, &state);
+
+  // XButtonEvent and XMotionEvent are almost identical, so we
+  // don't have to care which it is for these fields
+  dst->xbutton.serial = XLastKnownRequestProcessed(fl_display);
+  dst->xbutton.display = fl_display;
+  dst->xbutton.window = wnd;
+  dst->xbutton.root = root;
+  dst->xbutton.subwindow = None;
+  dst->xbutton.time = fl_event_time;
+  dst->xbutton.x = src.event_x;
+  dst->xbutton.y = src.event_y;
+  dst->xbutton.x_root = x_root;
+  dst->xbutton.y_root = y_root;
+  dst->xbutton.state = state.mods;
+  dst->xbutton.state |= ((state.ptr_buttons >> 1) & 0x1f) << 8;
+  dst->xbutton.same_screen = True; // FIXME
+}
+
+void XInputTouchHandler::fakeMotionEvent(const GHEvent origEvent)
+{
+  XEvent fakeEvent;
+  int mask;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = MotionNotify;
+  fakeEvent.xmotion.is_hint = False;
+  preparePointerEvent(&fakeEvent, origEvent);
+
+  fakeEvent.xbutton.state |= buttonmask;
+
+  pushFakeEvent(&fakeEvent);
+}
+
+void XInputTouchHandler::fakeButtonEvent(bool press, int button,
+                                         const GHEvent origEvent)
+{
+  XEvent fakeEvent;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = press ? ButtonPress : ButtonRelease;
+  fakeEvent.xbutton.button = button;
+  preparePointerEvent(&fakeEvent, origEvent);
+
+  switch (button) {
+  case 1: // left mousebutton
+    if (press) {
+      buttonmask |= Button1Mask;
+    } else {
+      buttonmask &= ~Button1Mask;
+    }
+    break;
+  case 2: // middle mousebutton
+    if (press) {
+      buttonmask |= Button2Mask;
+    } else {
+      buttonmask &= ~Button2Mask;
+    }
+    break;
+  case 3: // right mousebutton
+    if (press) {
+      buttonmask |= Button3Mask;
+    } else {
+      buttonmask &= ~Button3Mask;
+    }
+    break;
+  case 4: // scroll up
+    if (press) {
+      buttonmask |= Button4Mask;
+    } else {
+      buttonmask &= ~Button4Mask;
+    }
+    break;
+  case 5: // scroll down
+    if (press) {
+      buttonmask |= Button5Mask;
+    } else {
+      buttonmask &= ~Button5Mask;
+    }
+    break;
+  case 6: // scroll left
+    if (press) {
+      buttonmask |= Button6Mask;
+    } else {
+      buttonmask &= ~Button6Mask;
+    }
+    break;
+  case 7: // scroll right
+    if (press) {
+      buttonmask |= Button7Mask;
+    } else {
+      buttonmask &= ~Button7Mask;
+    }
+    break;
+
+  default:
+    vlog.error(_("Invalid mouse button %i, must be a number between 1 and 7."),
+               button);
```

Comment:
This seems incorrect. Don't impose VNC limitations on the X11 code.

Reply:
Removed as part of 29be66fbc9ac52f396aedcb59738d5efb5bf7e14